sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala

```diff
@@ -177,4 +177,18 @@ class CanonicalizeSuite extends SparkFunSuite {
     assert(expr.semanticEquals(attr))
     assert(attr.semanticEquals(expr))
   }
+
+  test("SPARK-38030: Canonicalize Cast should remove nullability of target dataType") {
+    val structType = StructType(Seq(StructField("name", StringType, nullable = false)))
+    val attr = AttributeReference("col", structType)()
+    for (cast <- Seq(
+        Cast(attr, structType),
+        AnsiCast(attr, structType),
+        TryCast(attr, structType))) {
+      assert(cast.resolved)
+      // canonicalization should not converted resolved cast to unresolved
+      assert(cast.canonicalized.resolved)
```

Comment:
should we be asserting on `cast.canonicalized.output` as well? this is how the original issue was detected, right?

Reply:
The original issue was detected because of the following plan `UnionExec -> ProjectExec --> Alias --> Cast --> AttributeReference` and a call to `.output` on `UnionExec` which in turn calls `ProjectExec.output.nullable`. I can recreate this chain here, but not sure if thats useful. The root issue was that a resolved node was being converted to unresolved which we are testing here.

Reply:
Yah I think the current test is probably sufficient.